internal/promotion/runner/builtin/tar_extractor.go

```diff
@@ -0,0 +1,336 @@
+		isDir := header.Typeflag == tar.TypeDir
+		pathParts := strings.Split(targetName, "/")
+		if matcher.Match(pathParts, isDir) {
+			logger.Debug("ignoring path based on pattern", "path", targetName)
+			continue
+		}
+
+		// Create destination directory for files and links
+		targetPath := filepath.Join(outPath, targetName)
+
+		// Double check the target path is within the output directory
+		relPath, err := filepath.Rel(outPath, targetPath)
+		if err != nil || strings.HasPrefix(relPath, "..") || strings.HasPrefix(relPath, "/") {
+			logger.Debug("skipping path escaping target directory", "path", targetName)
+			continue
+		}
+
+		destDir := filepath.Dir(targetPath)
+		if err := os.MkdirAll(destDir, 0755); err != nil {
+			return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+				fmt.Errorf("failed to create directory %s: %w", destDir, err)
+		}
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			// Create directory
+			if err := os.MkdirAll(targetPath, 0755); err != nil {
+				return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+					fmt.Errorf("failed to create directory %s: %w", targetPath, err)
+			}
+
+		case tar.TypeSymlink:
+			// Validate the symlink target to prevent symlink attacks
+			if !t.validRelPath(header.Linkname) {
+				logger.Debug("skipping symlink with unsafe target", "path", targetName, "target", header.Linkname)
+				continue
+			}
+			// Create symlink
+			if err := os.Symlink(header.Linkname, targetPath); err != nil && !os.IsExist(err) {
+				logger.Debug("failed to create symlink", "path", targetPath, "target", header.Linkname, "error", err)
+			}
+
+		case tar.TypeReg:
+			// Check if single file exceeds max size limit
+			logger.Debug("checking file size", "path", targetName, "size", header.Size)
+			if header.Size > MaxDecompressedFileSize {
+				logger.Debug("aborting extraction due to exceeding file size limit",
+					"path", targetName,
+					"size", header.Size,
+					"limit", MaxDecompressedFileSize)
+				return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+					fmt.Errorf("extraction aborted: file %s exceeds size limit of %d bytes", targetName, MaxDecompressedFileSize)
+			}
+
+			// Check if total extracted size would exceed limit
+			if totalExtractedSize+header.Size > MaxDecompressedTarSize {
+				logger.Debug("aborting extraction due to exceeding total size limit",
+					"total_size", totalExtractedSize,
+					"file_size", header.Size,
+					"limit", MaxDecompressedTarSize)
+				return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+					fmt.Errorf("extraction aborted: total size would exceed limit of %d bytes", MaxDecompressedTarSize)
+			}
+
+			// Create file
+			outFile, err := os.Create(targetPath)
+			if err != nil {
+				return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+					fmt.Errorf("failed to create file %s: %w", targetPath, err)
+			}
```

Comment:
The `os.Create` and `os.Chmod` can be combined as:

```go
os.OpenFile(targetPath, O_RDWR|O_CREATE|O_TRUNC, <permissions>)
```

I also do wonder if we should copy the modification data, as done here: https://cs.opensource.google/go/x/build/+/e235f28d:internal/untar/untar.go;l=114-123